### ISSUE TITLE
Removal of ingress networks when empty of other containers

### ISSUE THREAD

**Issue by ctelfer** [docker/libnetwork#2097]
This comes from deeper investigation an issue discovered by internal tests.  When swarmkit removes containers that are members of an ingress network, they can end up removing the ingress network itself when the last container leaves.   This is not intended behavior since an ingress network is supposed to stay in the swarm regardless.  The issue is simple to reproduce: 
  * create a "fresh" (no other containers, etc) swarm of 3 nodes:  1 manager and 2 workers
  * run `docker network ls` on the workers and observe that the ingress network is present
  * start a simple service on all 3 nodes.  For example: `docker service create -p80:80 --name nginx --replicas 3 nginx`
  * now remove the service
  * finally re-run `docker network ls` on the worker nodes.
The ingress network will no longer be present.

This is due to logic in the method network.delete(bool) in network.go:
```go
        if len(n.loadBalancerIP) != 0 {
                endpoints := n.Endpoints()
                if force || len(endpoints) == 1 {
                        n.deleteLoadBalancerSandbox()
                }
                //Reload the network from the store to update the epcnt.
                n, err = c.getNetworkFromStore(id)
                if err != nil {
                        return &UnknownNetworkError{name: name, id: id}
                }
        }
```
This comes right before the test to see if the network has no more endpoints.  Since ingress networks also have loadbalancer IP addresses, they will fall into this logic which will remove the remaining endpoint from the network and delete it.  This *should* happen for other dynamically created networks but not for ingress networks.  (i.e. networks for which n.ingress == true)

**Comment by ctelfer** [comment]
The obvious fix would be to change the test `len(endpoints) == 1` to `(len(endpoints) == 1 && !n.ingress` ...  This would make it so that ingress networks *never* get removed except as part of controller.networkCleanup() (store.go:482).   Not sure if this is what we really want for a solution though.

**Comment by ctelfer** [comment]
The problem with the above fix is that it prevents explicit removal of ingress networks which is a supported feature.   However, this can easily be remedied by re-activating some existing code in moby:
```
@@ -540,6 +540,10 @@ func (daemon *Daemon) deleteNetwork(nw libnetwork.Network, dynamic bool) error {
                return errdefs.Forbidden(err)
        }
 
+       if nw.Info().Ingress() {
+               d.deleteLoadBalancerSandbox(nw)
+       }
+
        if err := nw.Delete(); err != nil {
                return err
        }
```
It really seems like the logic for removing the load balancing sandbox should be part of libnetwork and not moby.  However, this is probably the least-intrusive patch to fix the bug for the time being.  Refactoring can happen with other load-balancing modifications that are in progress.

**Comment by ctelfer** [comment]
Further examination and experimentation with the moby/moby source code reveals that removal of the ingress network follows a different path than other network removals which includes invoking releaseIngress().  So a simple patch in said function to invoke the aforementioned daemon.deleteLoadBalancerSandbox() suffices to re-enable explicit removal alongside of the aforementioned patch to prevent its implicit removal.

**Comment by ctelfer** [comment]
I've used this moby patch in tests to successfully remove ingress networks with the libnetwork fix described above.  Without said patch, the removal *appears* to work, but the network never actually gets removed.
```
christopher:~/work/src/github.com/moby/moby$ git diff master...
diff --git a/daemon/network.go b/daemon/network.go
index 2c2a96b..093b074 100644
--- a/daemon/network.go
+++ b/daemon/network.go
@@ -212,6 +212,8 @@ func (daemon *Daemon) releaseIngress(id string) {
                return
        }
 
+       daemon.deleteLoadBalancerSandbox(n)
+
        if err := n.Delete(); err != nil {
                logrus.Errorf("Failed to delete ingress network %s: %v", n.ID(), err)
                return
```

**Comment by ctelfer** [comment]
A bit of further color: swarm also creates the docker_gwbridge network on every node.  Said networks are local in scope and have a veth pair attached to the ingress network.  When one removes the ingress network manually (when using the libnetwork and moby fixes) the system removes the docker_gwbridge networks from the worker presumably because they are neither pre-defined networks, nor an ingress network.  However it does not do this on the manager that one sends the removal command to.  Manually removing docker_gwbridge on said manager after removing ingress works just fine however.  This is a bit of quirky behavior.

**Comment by ctelfer** [comment]
I have a new theory w.r.t. https://github.com/docker/libnetwork/issues/2097 based on discussions w/ @ddebroy yesterday.  Maybe swarm is working exactly as designed.  It looks like whenever swarm is reconciling its new node state w/ its previous node state it takes a fresh look at the ingress networks.   If it sees that they are no longer needed it removes them.  If it sees that they are now needed (but weren't before) it tries to recreate them and removes any old stale entries.
In other words it may well intend to be removing the ingress sandbox and the fact that it only partially does on the manager is, in fact, the bug....
If this were true, however, then it raises a few other questions:
 * why have the swarm nodes spin up w/ ingress and docker_gwbridge in the first place if they will just get recreated when needed?
 * why don't the e2e tests work?  Is it some race between ingress creation/deletion and node state reconciliation during service spin-up?  Or maybe it has to do with the fact that the e2e tests are running within docker-in-docker and so don't communicate w/ swarm in the same way?  (don't know if this is even true, just postulating)
 * even if true, maybe it is still the swarm code that is mistaken in trying to remove said ingress network when it is "not needed"

Even if this theory is correct, the way moby interacts with libnetwork is buggy, because manager nodes delete ingress networks from the libnetwork store but do not seem to completely remove them from the cluster state from what we can tell.  This seems to cause things like `docker network ls` and `docker network inspect ingress` to still show the ingress network (albiet w/ corrupted values in the case of the latter), but w/ libnetwork clearly not having the network present.

**Comment by ctelfer** [comment]
Ok, have a full RCA for this at this point.   The first issue is the fact that when the reference count for ANY network drops down to 1 and its a load balancing endpoint, then the network gets removed from libnetwork.  This is clearly intended for dynamic networks with load-balancing sandboxes like those created by Swarm w/ Windows nodes.  However it is NOT intended for the ingress network and as a result removing the last "regular" endpoint from the ingress network causes libnetwork to remove the ingress network.   This is addressed by   https://github.com/docker/libnetwork/pull/2101 as mentioned.

However, the above fix will also *prevent* explicit removal of the ingress network because the current moby code for removing the ingress network relies on the aforementioned bug.  It didn't always rely on said bug.  The patch https://github.com/docker/libnetwork/pull/2011 moved the logic that was removing load-balancing endpoints (for non-ingress networks) from moby into libnetwork.   Unfortunately, the new location of that code also meant that the code *specifically lost context* for whether the network removal was explicit or implicit when the network happened to be ingress.  So the newer logic in libnetwork simply doesn't have the state information to determine whether to remove the load balancing endpoint.  It needs to know: whether there is such an endpoint (check), whether it is the last endpoint in the network (check), whether the network is an ingress network (check) and finally, if all of the above are true, whether the deletion was explicit or implicit (information not available).

So fixing this bug specifically requires either passing this additional information to the network.delete() function or making moby perform the endpoint removal for the ingress network.   The former costs a change in the API which seems highly undesirable.  The latter could take two forms:  moby could find and remove the endpoint itself or moby could invoke a new libnetwork function that does the removal.  Among these two choices, the latter seems a bit cleaner because, although it is an API change, it doesn't change the signatures of any existing methods and it keeps the logic within libnetwork.  On the other hand, the code to perform the explicit removal is already present in moby as mentioned above.  So the former change (explicit removal by moby) is very small.  Moreover, moby clearly already has a lot of special logic for managing the ingress network and so a bit extra here is not unusual.

A final important note regards why the ingress networks do not vanish completely from manager nodes due to this bug.   This is because the manager nodes in a swarm consult both the libnetwork store *and* the cluster network store when searching for networks at the behest of commands like `docker network ls` and `docker network inspect`.   Removing the last non-load-balancing endpoint from ingress does indeed cause libnetwork to stop tracking the network.  But it does not in *any way* affect the list of cluster networks in the swarm.   As a result, query commands like `inspect` still get information about the ingress network even though the actual sandboxes and interfaces are destroyed.   This makes writing a regression test for this use case a bit trickier.

**Comment by ctelfer** [comment]
Fixed and merged in moby/moby#36538.
